### PR TITLE
go checker

### LINKED ISSUE #21: Import rules from internet and test them
Now really important thing is to allow our application to download rules from internet and test whether they passes

### PR DESCRIPTION
**Pull request by BaderBC**
Closes #21

### CODE REVIEW

File: src/checker/check.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	if err != nil || !isValid {
+		panic("Rule url is invalid!")
+	}
+	if url[len(url)-1] == '/' {
+		url = url[:len(url)-1]
+	}
+	
+	if rulesHistory.Contains(url, name) {
+		if rulesHistory.IsRuleInProgress(url, name) {
+			// TODO: consider error handling
```
Comment (maxidragon) [inline src/checker/check.go:55]:
```suggestion

```

File: src/checker/check.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	rule := r[url+name]
+	rule.isInProgress = isInProgress
+}
+
+func CheckRuleScript(script string) (bool, error) {
+ 	rulesHistory := &RulesHistory{}
+	return ExecuteLuaMain(script, rulesHistory)
+}
+
+func CheckRule(rulesHistory *RulesHistory, url string, name string) bool {
```
Comment (cooligus) [inline src/checker/check.go:43]:
```go
func CheckRule(rulesHistory *RulesHistory, url string, name string) (bool, err) {
```

File: src/checker/check.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	const URL_REG_EXP = "https?:\\/\\/(www\\.)?[-a-zA-Z0-9@:%._\\+~#=]{1,256}\\.[a-zA-Z0-9()]{1,6}\\b([-a-zA-Z0-9()@:%_\\+.~#?&//=]*)"
+	isValid, err := regexp.MatchString(URL_REG_EXP, url)
+	if err != nil || !isValid {
+		panic("Rule url is invalid!")
```
Comment (cooligus) [inline src/checker/check.go:47]:
```go
return false, errors.New("rule url is invalid!")
```

File: src/checker/check.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	}
+	if url[len(url)-1] == '/' {
+		url = url[:len(url)-1]
+	}
+	
+	if rulesHistory.Contains(url, name) {
+		if rulesHistory.IsRuleInProgress(url, name) {
+			panic("ERROR: Dependencies creates infinity loop")
```
Comment (cooligus) [inline src/checker/check.go:55]:
```go
return false, errors.New("dependencies creates infinity loop")
```

File: src/checker/check.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+		} else {
+			return true
+		}
+	}
+	rulesHistory.Push(url, name, true)
+
+	err, ruleSetPath, removeTempDir := FetchRuleSet(url)
+	if err != nil {
+		println(err.Error())
+		panic("Failed to fetch rules from git repo: " + url)
```
Comment (cooligus) [inline src/checker/check.go:65]:
```go
return false, errors.New("failed to fetch rules from git repo: " + url)
```

File: src/checker/check.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+	}
+	defer removeTempDir()
+
+	script, err := getScript(ruleSetPath, name)
+	if err != nil {
+		println(err.Error())
+		panic("Failed to read script called: " + name + " in git repo: " + url)
```
Comment (cooligus) [inline src/checker/check.go:72]:
```go
return false, errors.New("failed to read script called: " + name + " in git repo: " + url)
```

**Review by cooligus — CHANGES_REQUESTED** [review]
Please don't use `panic`. Anyway, I think the way of storing rules is an issue that need to be discussed. Downloading and immediately deleting after use isn't good idea IMO. Moreover, when I saw function checking URL, I immediately thought of a library that should do this for us and should be available on the internet.

File: src/checker/check.go (hunk 2 above)
Comment (maxidragon) [reply src/checker/check.go:43]: @cooligus when you write code in a review please add it as suggestion
image

**Comment by maxidragon** [comment]
> Anyway, I think the way of storing rules is an issue that need to be discussed. 

I agree with you. We should discuss it at our next meeting

**Comment by BaderBC** [comment]
> Please don't use `panic`. Anyway, I think the way of storing rules is an issue that need to be discussed. Downloading and immediately deleting after use isn't good idea IMO. Moreover, when I saw function checking URL, I immediately thought of a library that should do this for us and should be available on the internet.

1) That was a draft, now everything looks different
2) Speaking about using panic, it's much more complicated than I thought. I cannot use error propagation because CheckRule function is directly invoked from lua. That's why I cannot propagate error higher that CheckRule.
I was thinking about calling some function which will send appropriate message to CLI / GUI and then exit with code 1. But for now we have neither CLI nor GUI

File: src/checker/check.go (hunk 2 above)
Comment (cooligus) [reply src/checker/check.go]: ok

File: src/checker/resolve_script.go [hunk 7]
```diff
@@ -1,13 +1,81 @@
...
+	val, ok := s.Rules[key].(RuleConf)
+	return val, ok
+}
+
+func getRulePath(ruleSetLocation RuleSetLocation, ruleName string) (string, error) {
+	// Support for both .yaml and .yml
+	spitoRulesDataBytes, err := os.ReadFile(ruleSetLocation.getRuleSetPath() + "/SPITO_RULES.yaml")
+	if errors.Is(err, fs.ErrNotExist) {
+		var _err error
+		spitoRulesDataBytes, _err = os.ReadFile(ruleSetLocation.getRuleSetPath() + "/SPITO_RULES.yml")
```
Comment (cooligus) [inline src/checker/resolve_script.go:35]: I suggest you to create constant with name of spito-rules file.

**Review by cooligus — COMMENTED** [review]
Can't you just propagate error to lua? Anyway, I think there are still issues that need to be discussed on our meeting.

File: src/checker/integration_tests/rule_require_test.go [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	doesRulePasses, err := checker.CheckRuleScript(string(script))
+	if err != nil {
+		t.Fatal(err)
+	}
+	
+	println("rule passes: ", doesRulePasses)
```
Comment (maxidragon) [inline src/checker/integration_tests/rule_require_test.go:20]:
```suggestion
	println("Rule passes: ", doesRulePasses)
```

**Review by maxidragon — COMMENTED** [review]
One nit. Please do not merge it before our next meeting, as I and @cooligus said we should discuss how we should store our rules.

**Review by sewe2000 — APPROVED** [review]

File: src/checker/integration_tests/rule_require_test.go [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+	"testing"
+)
+
+func TestRuleRequire(t *testing.T) {
+	script, err := os.ReadFile("./rule_require_test.lua")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	doesRulePasses, err := checker.CheckRuleScript(string(script))
```
Comment (sewe2000) [inline src/checker/integration_tests/rule_require_test.go:15]: I think 'doesRulePasses' variable should be renamed to 'doesRulePass'

**Review by sewe2000 — CHANGES_REQUESTED** [review]

File: src/checker/utils.go [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"strings"
+)
+
+const tmpRuleSetsDir = "/tmp/spito-rules/rule-sets"
```
Comment (cooligus) [inline src/checker/utils.go:11]: Can you change `/tmp` to `/home/${user}/.local/state`?

**Review by cooligus — APPROVED** [review]

**Pull request merged**